engine/src/main/java/org/hibernate/search/engine/backend/work/execution/impl/OperationSubmitterType.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.engine.backend.work.execution.impl;
+
+import java.lang.invoke.MethodHandles;
+import java.util.Queue;
+import java.util.concurrent.BlockingQueue;
+
+import org.hibernate.search.engine.backend.work.execution.spi.OperationSubmitter;
+import org.hibernate.search.engine.backend.work.execution.spi.WorkQueueFullException;
+import org.hibernate.search.engine.logging.impl.Log;
+import org.hibernate.search.util.common.logging.impl.LoggerFactory;
+
+public enum OperationSubmitterType implements OperationSubmitter {
+	/**
+	 * {@code OperationSubmitter} implementation that will block the process and wait for the operation to be submitted
+	 * to the queue.
+	 */
+	BLOCKING {
+		@Override
+		public <T> void submitToQueue(Queue<T> queue, T element) {
+			if ( queue instanceof BlockingQueue ) {
+				try {
+					( (BlockingQueue<T>) queue ).put( element );
+				}
+				catch (InterruptedException e) {
+					throw log.failedToSubmitToQueue( element, e.getMessage(), e );
+				}
```

Comment:
That's not how `InterruptedException` should be handled; you're not resetting the interrupt flag correctly.

Regardless, if you change the parameter type to `BlockingQueue`, it would make sense for the method to just let the `InterruptedException` go through and be caught by the caller if necessary.